File: two_pointers/three_sum.py
def three_sum_two_pointers(nums: list[int]) -> list[list[int]]:
    res = []
    nums.sort()
    for i, num in enumerate(nums):
        # Skip the same value to avoid duplicate triplets
        if i > 0 and num == nums[i - 1]:
            continue

        # Use two pointers for the remaining part of the array
        left, right = i + 1, len(nums) - 1
        while left < right:
            three_sum = num + nums[left] + nums[right]
            if three_sum > 0:
                right -= 1  # Sum too large, move right pointer left
            elif three_sum < 0:
                left += 1  # Sum too small, move left pointer right
            else:
                res.append([num, nums[left], nums[right]])
                left += 1
                right -= 1
                # Skip duplicates for the left pointer
                while nums[left] == nums[left - 1] and left < right:
                    left += 1
    return res


def three_sum_hashset(nums: list[int]) -> list[list[int]]:
    res = set()
    nums.sort()  # Sorting helps in handling duplicates easily

    for i in range(len(nums)):
        # Optimization: If the current number is > 0, no triplet can sum to 0
        if nums[i] > 0:
            break
        # Skip duplicate values for the first element
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        seen = set()
        for j in range(i + 1, len(nums)):
            complement = -nums[i] - nums[j]
            if complement in seen:
                # Found a triplet! Use a tuple so it can be added to the result set
                res.add((nums[i], complement, nums[j]))

                # To avoid duplicates in the inner loop, skip same values
                while j + 1 < len(nums) and nums[j] == nums[j + 1]:
                    j += 1
            seen.add(nums[j])

    return [list(triplet) for triplet in res]


def three_sum_hashset_no_sort(nums: list[int]) -> list[list[int]]:
    res = set()
    seen = set()

    for i in range(len(nums)):
        if nums[i] in seen:
            continue
        seen.add(nums[i])

        # Find Two Sum for the remainder of the array
        target = -nums[i]
        visited = set()

        for j in range(i + 1, len(nums)):
            complement = target - nums[j]
            if complement in visited:
                triplet = tuple(sorted([nums[i], nums[j], complement]))
                res.add(triplet)
            visited.add(nums[j])

    return [list(t) for t in res]

File: two_pointers/three_sum.ts
function threeSumTwoPointers(nums: number[]): number[][] {
    const res: number[][] = [];
    nums.sort((a, b) => a - b);

    for (let i = 0; i < nums.length; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue;

        let left = i + 1, right = nums.length - 1;
        while (left < right) {
            const sum = nums[i] + nums[left] + nums[right];
            if (sum > 0) right--;
            else if (sum < 0) left++;
            else {
                res.push([nums[i], nums[left], nums[right]]);
                left++;
                right--;
                while (left < right && nums[left] === nums[left - 1]) left++;
            }
        }
    }

    return res;
}

function threeSumHashset(nums: number[]): number[][] {
    const res = new Set<string>();
    nums.sort((a, b) => a - b);

    for (let i = 0; i < nums.length; i++) {
        if (nums[i] > 0) break;
        if (i > 0 && nums[i] === nums[i - 1]) continue;

        const seen = new Set<number>();
        for (let j = i + 1; j < nums.length; j++) {
            const complement = -nums[i] - nums[j];
            if (seen.has(complement)) {
                res.add(JSON.stringify([nums[i], complement, nums[j]]));
            }
            seen.add(nums[j]);
        }
    }

    return [...res].map(s => JSON.parse(s));
}
